src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9121,22 +9142,45 @@ private ForStatementSyntax ParseForStatement(SyntaxList<AttributeListSyntax> att
         {
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ForKeyword);
 
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfForStatementArgument;
+
             var forToken = this.EatToken(SyntaxKind.ForKeyword);
             var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+            var (variableDeclaration, initializers) = eatVariableDeclarationOrInitializers();
 
-            var saveTerm = _termState;
-            _termState |= TerminatorState.IsEndOfForStatementArgument;
+            // Pulled out as we need to track this when parsing incrementors to place skipped tokens.
+            SyntaxToken secondSemicolonToken;
+            var forStatement = _syntaxFactory.ForStatement(
+                attributes,
+                forToken,
+                openParen,
+                variableDeclaration,
+                initializers,
+                firstSemicolonToken: eatCommaOrSemicolon(),
+                condition: this.CurrentToken.Kind is not SyntaxKind.SemicolonToken and not SyntaxKind.CommaToken
+                    ? this.ParseExpressionCore()
+                    : null,
+                secondSemicolonToken = eatCommaOrSemicolon(),
+                // Do allow semicolons (with diagnostics) in the incrementors list.  This allows us to consume
+                // accidental extra incrementors that should have been separated by commas.
+                incrementors: this.CurrentToken.Kind != SyntaxKind.CloseParenToken
+                    ? parseForStatementExpressionList(ref secondSemicolonToken, allowSemicolonAsSeparator: true)
+                    : default,
+                eatUnexpectedTokensAndCloseParenToken(),
+                ParseEmbeddedStatement());
```

Comment:
broke parsing up into distinct parts for each major chunk of hte for-loop.  Teh variables/initializers, the semicolon, the condition, the semicolon, the incrementors, the close paren and the statement.